<a href="https://colab.research.google.com/github/JuanRoccia/MachineLearningUPSO/blob/main/Laboratorio_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***Primera Parte***
*Caraga, Limpieza, Preparación, Analisis de los datos y Visualización*

### **1° Carga de datos:**
- Cargar los datos de ventas de cada archivo CSV.

In [ ]:
# Importando las bibliotecas necesarias
import os
import pandas as pd

# Definiendo la ruta donde se encuentran los archivos CSV
ruta_de_acceso = "/content/drive/MyDrive/ML&BD/Dataset_ventas2"

# Listando todos los archivos en el directorio especificado
archivos = os.listdir(ruta_de_acceso)

# Creando un diccionario vacío para almacenar los DataFrames de cada mes
dataframes = {}

# Iterando sobre cada archivo en la lista de archivos
for archivo in archivos:
    # Extrae el nombre del mes del nombre del archivo
    mes = archivo.split('_')[3].split('.')[0]

    # Cargando el archivo CSV en un DataFrame y almacenándolo en el diccionario con el nombre del mes como clave
    dataframes[mes] = pd.read_csv(os.path.join(ruta_de_acceso, archivo))

# Verificando la carga de datos mostrando las primeras filas del DataFrame de Enero
dataframes['Enero'].head()


ID de Pedido                  Producto Cantidad Pedida Precio Unitario  \
0       141234                    iPhone               1             700   
1       141235  Lightning Charging Cable               1           14.95   
2       141236          Wired Headphones               2           11.99   
3       141237          27in FHD Monitor               1          149.99   
4       141238          Wired Headphones               1           11.99   

  Fecha de Pedido                     Dirección de Envio  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

In [ ]:
# 1. Antes de la limpieza
filas_antes = {mes: len(df) for mes, df in dataframes.items()}

### **2° Limpieza de los datos:**
- Eliminar datos no numéricos.
- Eliminar filas incompletas

In [ ]:
for mes, df in dataframes.items():
    # Convertir las columnas 'Cantidad Pedida' y 'Precio Unitario' a numéricas y reemplazar los valores no numéricos con NaN
    df['Cantidad Pedida'] = pd.to_numeric(df['Cantidad Pedida'], errors='coerce')
    df['Precio Unitario'] = pd.to_numeric(df['Precio Unitario'], errors='coerce')

    # Eliminar filas incompletas
    df.dropna(subset=['Cantidad Pedida', 'Precio Unitario'], inplace=True)

    # Actualizar el DataFrame en el diccionario
    dataframes[mes] = df


# Convertir 'Cantidad Pedida' a entero
df['Cantidad Pedida'] = df['Cantidad Pedida'].astype(int)
df

ID de Pedido                Producto  Cantidad Pedida  Precio Unitario  \
0           248151   AA Batteries (4-pack)                4             3.84   
1           248152    USB-C Charging Cable                2            11.95   
2           248153    USB-C Charging Cable                1            11.95   
3           248154        27in FHD Monitor                1           149.99   
4           248155    USB-C Charging Cable                1            11.95   
...            ...                     ...              ...              ...   
11681       259353  AAA Batteries (4-pack)                3             2.99   
11682       259354                  iPhone                1           700.00   
11683       259355                  iPhone                1           700.00   
11684       259356  34in Ultrawide Monitor                1           379.99   
11685       259357    USB-C Charging Cable                1            11.95   

      Fecha de Pedido                       Dirección de Envio  
0      09/17/19 14:44      380 North St, Los Angeles, CA 90001  
1      09/29/19 10:19             511 8th St, Austin, TX 73301  
2      09/16/19 17:48    151 Johnson St, Los Angeles, CA 90001  
3      09/27/19 07:52        355 Hickory St, Seattle, WA 98101  
4      09/01/19 19:03            125 5th St, Atlanta, GA 30301  
...               ...                                      ...  
11681  09/17/19 20:56   840 Highland St, Los Angeles, CA 90001  
11682  09/01/19 16:00  216 Dogwood St, San Francisco, CA 94016  
11683  09/23/19 07:39     220 12th St, San Francisco, CA 94016  
11684  09/19/19 17:30   511 Forest St, San Francisco, CA 94016  
11685  09/30/19 00:18   250 Meadow St, San Francisco, CA 94016  

[11629 rows x 6 columns]

In [ ]:
# 2. Después de la limpieza
filas_despues = {mes: len(df) for mes, df in dataframes.items()}

# Mostrar la comparación
for mes in dataframes.keys():
    print(f"Mes: {mes}")
    print(f"Filas antes de la limpieza: {filas_antes[mes]}")
    print(f"Filas después de la limpieza: {filas_despues[mes]}")
    print("------")

Mes: Abril
Filas antes de la limpieza: 18383
Filas después de la limpieza: 18289
------
Mes: Agosto
Filas antes de la limpieza: 12011
Filas después de la limpieza: 11957
------
Mes: Diciembre
Filas antes de la limpieza: 25117
Filas después de la limpieza: 24989
------
Mes: Enero
Filas antes de la limpieza: 9723
Filas después de la limpieza: 9681
------
Mes: Febrero
Filas antes de la limpieza: 12036
Filas después de la limpieza: 11986
------
Mes: Julio
Filas antes de la limpieza: 14371
Filas después de la limpieza: 14291
------
Mes: Junio
Filas antes de la limpieza: 13622
Filas después de la limpieza: 13556
------
Mes: Marzo
Filas antes de la limpieza: 15226
Filas después de la limpieza: 15154
------
Mes: Mayo
Filas antes de la limpieza: 16635
Filas después de la limpieza: 16554
------
Mes: Noviembre
Filas antes de la limpieza: 17661
Filas después de la limpieza: 17580
------
Mes: Octubre
Filas antes de la limpieza: 20379
Filas después de la limpieza: 20284
------
Mes: Septiembre
Filas 

### **3° Preparación de los datos:**
- Ajustar los tipos de datos de cada atributo
- Extraer características importantes para el análisis como meses, horas, ciudades etc

In [ ]:
for mes, df in dataframes.items():
    # Convertir 'Fecha de Pedido' a datetime y extraer mes, día, hora y año:
    df['Fecha de Pedido'] = pd.to_datetime(df['Fecha de Pedido'], format='%m/%d/%y %H:%M')
    df['Año'] = df['Fecha de Pedido'].dt.year
    df['Mes'] = df['Fecha de Pedido'].dt.month
    df['Dia'] = df['Fecha de Pedido'].dt.day
    df['Hora'] = df['Fecha de Pedido'].dt.hour

    # Extraer ciudad y estado de 'Dirección de Envio':
    df['Ciudad'] = df['Dirección de Envio'].apply(lambda x: x.split(',')[1].strip())
    df['Estado'] = df['Dirección de Envio'].apply(lambda x: x.split(',')[2].split()[0])

    # Actualizar el DataFrame en el diccionario
    dataframes[mes] = df

df

ID de Pedido                Producto  Cantidad Pedida  Precio Unitario  \
0           248151   AA Batteries (4-pack)                4             3.84   
1           248152    USB-C Charging Cable                2            11.95   
2           248153    USB-C Charging Cable                1            11.95   
3           248154        27in FHD Monitor                1           149.99   
4           248155    USB-C Charging Cable                1            11.95   
...            ...                     ...              ...              ...   
11681       259353  AAA Batteries (4-pack)                3             2.99   
11682       259354                  iPhone                1           700.00   
11683       259355                  iPhone                1           700.00   
11684       259356  34in Ultrawide Monitor                1           379.99   
11685       259357    USB-C Charging Cable                1            11.95   

          Fecha de Pedido                       Dirección de Envio   Año  Mes  \
0     2019-09-17 14:44:00      380 North St, Los Angeles, CA 90001  2019    9   
1     2019-09-29 10:19:00             511 8th St, Austin, TX 73301  2019    9   
2     2019-09-16 17:48:00    151 Johnson St, Los Angeles, CA 90001  2019    9   
3     2019-09-27 07:52:00        355 Hickory St, Seattle, WA 98101  2019    9   
4     2019-09-01 19:03:00            125 5th St, Atlanta, GA 30301  2019    9   
...                   ...                                      ...   ...  ...   
11681 2019-09-17 20:56:00   840 Highland St, Los Angeles, CA 90001  2019    9   
11682 2019-09-01 16:00:00  216 Dogwood St, San Francisco, CA 94016  2019    9   
11683 2019-09-23 07:39:00     220 12th St, San Francisco, CA 94016  2019    9   
11684 2019-09-19 17:30:00   511 Forest St, San Francisco, CA 94016  2019    9   
11685 2019-09-30 00:18:00   250 Meadow St, San Francisco, CA 94016  2019    9   

       Dia  Hora         Ciudad Estado  
0       17    14    Los Angeles     CA  
1       29    10         Austin     TX  
2       16    17    Los Angeles     CA  
3       27     7        Seattle     WA  
4        1    19        Atlanta     GA  
...    ...   ...            ...    ...  
11681   17    20    Los Angeles     CA  
11682    1    16  San Francisco     CA  
11683   23     7  San Francisco     CA  
11684   19    17  San Francisco     CA  
11685   30     0  San Francisco     CA  

[11629 rows x 12 columns]

### **4° Análisis de datos, ejemplo:**

- Examinar las horas del día en que las ventas son más frecuentes.
- Determinar el mejor momento para mostrar publicidad.

In [ ]:
import pandas as pd
import plotly.graph_objects as go

# Lista de meses
meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']

# Combinar todos los DataFrames mensuales en uno solo
df_total = pd.concat(dataframes.values(), ignore_index=True)

# Crear una función para extraer la hora de la columna 'Fecha de Pedido'
def extraer_hora(fecha):
    return fecha.hour

# Aplicar la función para crear una nueva columna 'Hora'
df_total['Hora'] = df_total['Fecha de Pedido'].apply(extraer_hora)

# Crear una figura vacía
fig = go.Figure()

# Para cada mes, agregar una línea al gráfico que represente las unidades vendidas por hora
for mes in meses:
    df_mes = dataframes[mes]
    df_mes['Hora'] = df_mes['Fecha de Pedido'].apply(extraer_hora)
    unidades_vendidas_por_hora = df_mes.groupby('Hora')['Cantidad Pedida'].sum()
    # Convertir 'Cantidad Pedida' a entero
    df_total['Cantidad Pedida'] = df_total['Cantidad Pedida'].astype(int)
    fig.add_trace(go.Scatter(x=unidades_vendidas_por_hora.index, y=unidades_vendidas_por_hora.values, mode='lines', name=mes))

# Configurar el layout del gráfico
fig.update_layout(title='Unidades vendidas por hora de cada mes',
                   xaxis_title='Hora del día',
                   yaxis_title='Unidades vendidas',
                   xaxis=dict(tickvals=list(range(0,24))))

# Mostrar el gráfico
fig.show()


In [ ]:
import pandas as pd
import plotly.graph_objects as go

# Lista de meses
meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']

# Combinar todos los DataFrames mensuales en uno solo
df_total = pd.concat(dataframes.values(), ignore_index=True)

# Crear una función para extraer la hora de la columna 'Fecha de Pedido'
def extraer_hora(fecha):
    return fecha.hour

# Aplicar la función para crear una nueva columna 'Hora'
df_total['Hora'] = df_total['Fecha de Pedido'].apply(extraer_hora)

# Calcular las ventas para cada entrada
df_total['Ventas'] = df_total['Cantidad Pedida'] * df_total['Precio Unitario']

# Crear una figura vacía
fig = go.Figure()

# Para cada mes, agregar una línea al gráfico que represente las ventas por hora
for mes in meses:
    df_mes = dataframes[mes]
    df_mes['Hora'] = df_mes['Fecha de Pedido'].apply(extraer_hora)
    df_mes['Ventas'] = df_mes['Cantidad Pedida'] * df_mes['Precio Unitario']
    # Convertir 'Cantidad Pedida' a entero
    df_total['Cantidad Pedida'] = df_total['Cantidad Pedida'].astype(int)
    ventas_por_hora_mes = df_mes.groupby('Hora')['Ventas'].sum()
    fig.add_trace(go.Scatter(x=ventas_por_hora_mes.index, y=ventas_por_hora_mes.values, mode='lines', name=mes))

# Configurar el layout del gráfico
fig.update_layout(title='Ventas totales por hora de cada mes',
                   xaxis_title='Hora del día',
                   yaxis_title='Ventas totales ($)',
                   xaxis=dict(tickvals=list(range(0,24))))

# Mostrar el gráfico
fig.show()


### **5° Visualización de datos:**
- Utilizar gráficos y visualizaciones para representar los resultados de los análisis.
- Crear gráficos de barras, líneas, mapas, tablas, etc, según corresponda.

In [ ]:
def extraer_datos(datos):
  datos['Fecha de Pedido'] = pd.to_datetime(datos['Fecha de Pedido'])
  datos['Hora'] = datos['Fecha de Pedido'].dt.hour
  datos['Dia'] = datos['Fecha de Pedido'].dt.day
  return datos

import plotly.graph_objects as go

meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']

totales_ventas = []
nombres_meses = []

for mes in meses:
  dataframe_mes = dataframes[mes]
  dataframes_mes = extraer_datos(dataframe_mes)

  total_ventas_mes = dataframe_mes['Cantidad Pedida'].sum()
  totales_ventas.append(total_ventas_mes)
  nombres_meses.append(mes)

fig = go.Figure()

# Agrego al objeto figura un grafico de barra conn los valores extraidos previamente
fig.add_trace(go.Bar(
        x= nombres_meses,
        y=  totales_ventas,
        yaxis="y",
        name="Unidades vendidas",
        marker = dict(color= totales_ventas, colorscale = 'RdYlGn',colorbar = dict(title = "Total de ventas")),
        text = totales_ventas,
        textposition = 'auto'
))

# Configuro titulos y ejes del grafico
fig.update_layout(
    title="Unidades vendidas por mes",
    xaxis=dict(title="Meses"),
    yaxis=dict(title="Unidades vendidas", side="left"))


# Muestro el grafico
fig.show()

# ***Segunda Parte***
*Preguntas a responder en el análisis*

### **1. Comportamiento de las ventas en los distintos meses:**
- ¿Cómo variaron las ventas a lo largo de los diferentes meses?
- ¿Hubo algún mes que se destacó en términos de ventas?
- ¿Cuál es el ingreso total generado por mes?

In [ ]:
import plotly.graph_objects as go

meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']

totales_ventas = []
nombres_meses = []

for mes in meses:
  dataframe_mes = dataframes[mes]
  dataframes_mes = extraer_datos(dataframe_mes)

  total_ventas_mes = dataframe_mes['Cantidad Pedida'].sum()
  totales_ventas.append(total_ventas_mes)
  nombres_meses.append(mes)

fig = go.Figure()

# Agrego al objeto figura un grafico de barra conn los valores extraidos previamente
fig.add_trace(go.Bar(
        x= nombres_meses,
        y=  totales_ventas,
        yaxis="y",
        name="Unidades vendidas",
        marker = dict(color= totales_ventas, colorscale = 'RdYlGn',colorbar = dict(title = "Total de ventas")),
        text = totales_ventas,
        textposition = 'auto'
))

# Configuro titulos y ejes del grafico
fig.update_layout(
    title="Unidades vendidas por mes",
    xaxis=dict(title="Meses"),
    yaxis=dict(title="Unidades vendidas", side="left"))


# Muestro el grafico
fig.show()

In [ ]:
import plotly.graph_objects as go

# Calcular el ingreso total por mes
ingreso_por_mes = df_total.groupby('Mes')['Ventas'].sum()

# Crear el gráfico
fig = go.Figure()

# Escala de colores basada en el valor del ingreso
colors = ['rgba(204,204,204,1)' if val < max(ingreso_por_mes) and val > min(ingreso_por_mes) else 'rgba(222,45,38,0.8)' for val in ingreso_por_mes]

fig.add_trace(go.Bar(
    x=meses,
    y=ingreso_por_mes,
    marker_color=colors,
    text=ingreso_por_mes,
    textposition='outside'
))

# Anotaciones para destacar el mes con el ingreso más alto y el mes con el ingreso más bajo
fig.update_layout(
    annotations=[
        dict(x=meses[ingreso_por_mes.idxmax() - 1], y=max(ingreso_por_mes), text='Máximo ingreso', arrowhead=4, ax=0, ay=-40),
        dict(x=meses[ingreso_por_mes.idxmin() - 1], y=min(ingreso_por_mes), text='Mínimo ingreso', arrowhead=4, ax=0, ay=40)
    ],
    title='Ingreso total por mes',
    xaxis_title='Mes',
    yaxis_title='Ingreso total ($)',
    xaxis=dict(tickvals=list(range(0,13)), ticktext=meses),
    template='plotly_white'  # Estilo limpio y profesional
)

fig.show()


> Los gráficos de barras muestran las unidades vendidas por mes y los ingresos. Según los resultados, se observa una tendencia clara en las ventas.

- Inicio del año (Enero - Marzo): Hay un aumento gradual en las ventas, comenzando con el mes de enero que tiene las ventas más bajas y aumentando hasta marzo.

- Medio del año (Abril - Septiembre): Las ventas alcanzan su punto máximo en abril y luego comienzan a disminuir, alcanzando un punto bajo en septiembre.

- Final del año (Octubre - Diciembre): Las ventas vuelven a aumentar en octubre, alcanzando su punto máximo en diciembre.

<h1>Análisis:</h1>

Enero suele ser un mes de ventas más bajas en muchas industrias debido a que
las personas gastan mucho en las festividades de diciembre y luego reducen su consumo en enero.

Por eso tiene sentido que Diciembre muestre el número más alto de unidades vendidas e ingresos.

Octubre también muestra un aumento significativo en las ventas, lo que podría deberse a eventos o festividades específicas de ese mes o a la anticipación de las festividades de fin de año.

<h1>Respuestas a las preguntas: </h1>

- **¿Cómo variaron las ventas a lo largo de los diferentes meses?**
 - Las ventas comenzaron bajas en enero, aumentaron gradualmente hasta marzo, alcanzaron su punto máximo en abril, disminuyeron hasta septiembre, y luego volvieron a aumentar, alcanzando su punto más alto en diciembre.

- **¿Hubo algún mes que se destacó en términos de ventas?**
 - Diciembre se destacó con las ventas más altas, seguido de cerca por octubre y abril.

- **¿Cuál es el ingreso total generado por mes?**
  - Enero: 1,822,257
  - Febrero: 2,202,022
  - Marzo: 2,807,100
  - Abril: 3,390,670
  - Mayo: 3,152,607
  - Junio: 2,577,802
  - Julio: 2,647,776
  - Agosto: 2,244,468
  - Septiembre: 2,097,560
  - Octubre: 3,736,727
  - Noviembre: 3,199,603
  - Diciembre: 4,613,443

  <h1>Conclusión:</h1>

  > Estos resultados proporcionan una visión clara de cómo se comportaron las ventas a lo largo del año y ayudan a tomar decisiones informadas sobre estrategias de marketing, inventario y otros aspectos relacionados con las ventas. Por ejemplo, se podría considerar aumentar el inventario o las promociones en los meses que tienden a tener mayores ventas, como octubre y diciembre.

### **2. Optimización de la publicidad y patrón de ventas por hora:**
- ¿Cuál es el momento ideal del día para presentar la publicidad y aumentar la
probabilidad de compra?
- ¿Cómo cambian los patrones de ventas por hora a lo largo del año?
- ¿Hay modificaciones en los patrones de ventas durante las horas de mayor actividad
en los distintos meses?


In [ ]:
import plotly.graph_objects as go

# Agrupar por hora y sumar las ventas
ventas_por_hora = df_total.groupby('Hora')['Ventas'].sum()

# Crear el gráfico
fig = go.Figure()

fig.add_trace(go.Bar(
    x=ventas_por_hora.index,
    y=ventas_por_hora.values,
    marker_color='rgb(26, 118, 255)'
))

fig.update_layout(
    title='Ventas totales por hora durante todo el año',
    xaxis_title='Hora del día',
    yaxis_title='Ventas totales ($)',
    xaxis=dict(tickvals=list(range(0,24)))
)

fig.show()


In [ ]:
import pandas as pd
import plotly.graph_objects as go

# Lista de meses
meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']

# Combinar todos los DataFrames mensuales en uno solo
df_total = pd.concat(dataframes.values(), ignore_index=True)

# Crear una función para extraer la hora de la columna 'Fecha de Pedido'
def extraer_hora(fecha):
    return fecha.hour

# Aplicar la función para crear una nueva columna 'Hora'
df_total['Hora'] = df_total['Fecha de Pedido'].apply(extraer_hora)

# Calcular las ventas para cada entrada
df_total['Ventas'] = df_total['Cantidad Pedida'] * df_total['Precio Unitario']

# Crear una figura vacía
fig = go.Figure()

# Para cada mes, agregar una línea al gráfico que represente las ventas por hora
for mes in meses:
    df_mes = dataframes[mes]
    df_mes['Hora'] = df_mes['Fecha de Pedido'].apply(extraer_hora)
    df_mes['Ventas'] = df_mes['Cantidad Pedida'] * df_mes['Precio Unitario']
    # Convertir 'Cantidad Pedida' a entero
    df_total['Cantidad Pedida'] = df_total['Cantidad Pedida'].astype(int)
    ventas_por_hora_mes = df_mes.groupby('Hora')['Ventas'].sum()
    fig.add_trace(go.Scatter(x=ventas_por_hora_mes.index, y=ventas_por_hora_mes.values, mode='lines', name=mes))

# Configurar el layout del gráfico
fig.update_layout(title='Ventas totales por hora de cada mes',
                   xaxis_title='Hora del día',
                   yaxis_title='Ventas totales ($)',
                   xaxis=dict(tickvals=list(range(0,24))))

# Mostrar el gráfico
fig.show()


In [ ]:
# Imprimir los datos en formato tabular
print("Hora del día |", " | ".join(meses))
print("-" * 100)

for hora in range(24):
    ventas_hora = []
    for mes in meses:
        df_mes = dataframes[mes]
        df_mes['Hora'] = df_mes['Fecha de Pedido'].apply(extraer_hora)
        df_mes['Ventas'] = df_mes['Cantidad Pedida'] * df_mes['Precio Unitario']
        ventas_por_hora_mes = df_mes.groupby('Hora')['Ventas'].sum()
        ventas_hora.append(str(round(ventas_por_hora_mes.get(hora, 0), 2)))
    print(f"{hora:12} |", " | ".join(ventas_hora))


Hora del día | Enero | Febrero | Marzo | Abril | Mayo | Junio | Julio | Agosto | Septiembre | Octubre | Noviembre | Diciembre
----------------------------------------------------------------------------------------------------
           0 | 40085.18 | 45686.3 | 55393.39 | 74329.16 | 64838.0 | 62697.91 | 46521.09 | 37634.35 | 51637.96 | 77076.82 | 63138.35 | 94682.76
           1 | 24052.69 | 29937.52 | 35948.58 | 48011.05 | 42252.89 | 31479.47 | 35967.36 | 24399.37 | 28032.67 | 53687.29 | 42242.02 | 64855.97
           2 | 15339.87 | 11896.24 | 19382.88 | 22384.2 | 16498.13 | 20624.72 | 10092.67 | 15690.51 | 21438.16 | 28874.98 | 18293.6 | 34335.48
           3 | 5738.84 | 9077.09 | 8469.49 | 19552.22 | 12024.07 | 12422.84 | 10235.04 | 16270.97 | 4856.34 | 17013.2 | 12473.14 | 17624.65
           4 | 6156.37 | 8646.9 | 9303.14 | 18628.47 | 17642.07 | 15946.32 | 13424.88 | 10555.92 | 10272.65 | 21231.23 | 11061.99 | 19791.07
           5 | 7535.75 | 16777.9 | 17616.03 | 20514.55 | 2247

<h1>Análisis:</h1>

Los datos presentados muestran las ventas por hora para cada mes del año. A partir de estos datos, se pueden observar patrones claros en las ventas y cómo varían a lo largo del día y entre los meses.

- **Patrón Diario:** Las ventas tienden a ser más bajas durante las primeras horas del día (0:00 a 6:00) y comienzan a aumentar a partir de las 7:00, alcanzando su punto máximo alrededor de las 12:00. Luego, las ventas se mantienen relativamente altas durante la tarde y comienzan a disminuir después de las 20:00.

- **Variaciones Mensuales:** Diciembre muestra las ventas más altas, especialmente durante la tarde/noche, lo que podría estar relacionado con las compras navideñas. Enero, por otro lado, tiene ventas más bajas, lo que es comprensible ya que muchas personas reducen su gasto después de las festividades de diciembre. Otros meses, como octubre, también muestran un aumento en las ventas, lo que podría estar relacionado con eventos o festividades específicas.

<h1>Respuestas a las preguntas: </h1>

- **¿Cuál es el momento ideal del día para presentar la publicidad y aumentar la probabilidad de compra?**

 - El momento ideal del día para presentar publicidad sería entre las 11:00 y las 20:00, con un pico especial alrededor de las 12:00 y las 19:00. Estas horas muestran las ventas más altas en la mayoría de los meses, lo que indica que los clientes están más activos y dispuestos a comprar durante este período.

- **¿Cómo cambian los patrones de ventas por hora a lo largo del año?**

 - Los patrones de ventas por hora varían a lo largo del año, con picos más pronunciados en ciertos meses como diciembre. Sin embargo, el patrón diario general de ventas bajas en la mañana y altas en la tarde se mantiene constante a lo largo del año.

- **¿Hay modificaciones en los patrones de ventas durante las horas de mayor actividad en los distintos meses?**

 - Sí, hay variaciones en los patrones de ventas durante las horas de mayor actividad en diferentes meses. Por ejemplo, mientras que diciembre muestra un pico pronunciado en ventas alrededor de las 19:00, otros meses como enero y febrero muestran picos más suaves durante las mismas horas.

<h1>Conclusión:</h1>

> Para optimizar la publicidad y las ventas, es esencial considerar estos patrones y adaptar las estrategias de marketing en consecuencia. Presentar anuncios durante las horas pico puede aumentar la visibilidad y la probabilidad de compra. Además, tener en cuenta las variaciones mensuales puede ayudar a planificar campañas de marketing específicas para ciertos meses, aprovechando los patrones de compra de los clientes.



### **3. Distribución de ventas por ubicación:**
- ¿En qué ciudades se han registrado las mayores ventas?
- ¿Cómo se comparan las ventas en diferentes estados o regiones?
- ¿Existe variación en las ventas por estado a lo largo de los meses?

In [ ]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.preprocessing import MinMaxScaler


# Coordenadas predefinidas para las ciudades en tu DataFrame
coordenadas = {
    "Boston": (42.3601, -71.0589),
    "Los Angeles": (34.0522, -118.2437),
    "San Francisco": (37.7749, -122.4194),
    "Dallas": (32.7767, -96.7970),
    "Seattle": (47.6062, -122.3321),
    "Atlanta": (33.7490, -84.3880),
    "New York City": (40.7128, -74.0060),
    "Portland": (45.5122, -122.6587),
    "Austin": (30.2672, -97.7431),
}

# Agrupar por ciudad y sumar las ventas
ventas_por_ciudad = df_total.groupby('Ciudad')['Ventas'].sum().reset_index()

# Añadir las coordenadas de latitud y longitud al DataFrame
ventas_por_ciudad['Latitud'] = ventas_por_ciudad['Ciudad'].map(lambda x: coordenadas.get(x, (None, None))[0])
ventas_por_ciudad['Longitud'] = ventas_por_ciudad['Ciudad'].map(lambda x: coordenadas.get(x, (None, None))[1])

# Escalar las ventas a un rango de tamaños de burbuja
scaler = MinMaxScaler(feature_range=(10, 50))  # El rango (10, 50) es solo un ejemplo, puedes ajustarlo según tus necesidades
ventas_por_ciudad['Ventas_escaladas'] = scaler.fit_transform(ventas_por_ciudad[['Ventas']])

# Crear subplots: 1 fila, 2 columnas
fig = make_subplots(
    rows=1, cols=2,
    column_widths=[0.3, 0.9],
    subplot_titles=('Ventas por Ciudad', 'Mapa de Ventas'),
    specs=[[{"type": "xy"}, {"type": "geo"}]]  # Especificar el tipo de subplot
)

# Añadir gráfico de barras a la primera columna
fig.add_trace(
    go.Bar(
        y=ventas_por_ciudad['Ciudad'],
        x=ventas_por_ciudad['Ventas'],
        orientation='h',
        marker=dict(color=ventas_por_ciudad['Ventas'], colorscale='Cividis')
    ),
    row=1, col=1
)

# Añadir mapa a la segunda columna
fig.add_trace(
    go.Scattergeo(
        lon=ventas_por_ciudad['Longitud'],
        lat=ventas_por_ciudad['Latitud'],
        text=ventas_por_ciudad['Ciudad'] + ": " + ventas_por_ciudad['Ventas'].astype(str),
        marker=dict(
            color=ventas_por_ciudad['Ventas'],
            colorscale='Cividis',
            size=ventas_por_ciudad['Ventas_escaladas'],
            sizemode='diameter',
            showscale=True,
            colorbar=dict(y=0.4)
        ),
    ),
    row=1, col=2
)

# Configurar el layout del gráfico
fig.update_layout(
    title_text='Ventas por Ciudad en Estados Unidos',
    title_x=0.5,  # Centrar el título
    geo=dict(scope='usa', domain=dict(x=[0.3, 1], y=[0, 1])),  # Ajustar el dominio del subplot geo
    width=1200,
    height=600
)

fig.show()


In [ ]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.preprocessing import MinMaxScaler

# Coordenadas predefinidas para las ciudades
coordenadas = {
    "Boston, MA": (42.3601, -71.0589),
    "Los Angeles, CA": (34.0522, -118.2437),
    "San Francisco, CA": (37.7749, -122.4194),
    "Dallas, TX": (32.7767, -96.7970),
    "Seattle, WA": (47.6062, -122.3321),
    "Atlanta, GA": (33.7490, -84.3880),
    "New York City, NY": (40.7128, -74.0060),
    "Portland, OR": (45.5122, -122.6587),
    "Portland, ME": (43.6615, -70.2553),
    "Austin, TX": (30.2672, -97.7431)
}

# Crear columna 'Ciudad-Estado'
df_total['Ciudad-Estado'] = df_total['Ciudad'] + ", " + df_total['Estado']

# Agrupar por 'Ciudad-Estado' y sumar las ventas
ventas_por_ciudad = df_total.groupby('Ciudad-Estado')['Ventas'].sum().reset_index()

# Añadir coordenadas al DataFrame
ventas_por_ciudad['Latitud'] = ventas_por_ciudad['Ciudad-Estado'].map(lambda x: coordenadas.get(x, (None, None))[0])
ventas_por_ciudad['Longitud'] = ventas_por_ciudad['Ciudad-Estado'].map(lambda x: coordenadas.get(x, (None, None))[1])

# Escalar las ventas de las ciudades
scaler = MinMaxScaler(feature_range=(10, 50))
ventas_por_ciudad['Ventas_escaladas'] = scaler.fit_transform(ventas_por_ciudad[['Ventas']])

# Agrupar por estado y sumar las ventas
ventas_por_estado = df_total.groupby('Estado')['Ventas'].sum().reset_index()

# Crear subplots
fig = make_subplots(
    rows=1, cols=2,
    column_widths=[0.3, 0.9],
    subplot_titles=('Ventas por Estado', 'Mapa de Ventas'),
    specs=[[{"type": "xy"}, {"type": "geo"}]]
)

# Añadir gráfico de barras
fig.add_trace(
    go.Bar(
        y=ventas_por_estado['Estado'],
        x=ventas_por_estado['Ventas'],
        orientation='h',
        marker=dict(color=ventas_por_estado['Ventas'], colorscale='Cividis')
    ),
    row=1, col=1
)

# Añadir mapa coroplético
fig.add_trace(go.Choropleth(
    locations=ventas_por_estado['Estado'],
    z=ventas_por_estado['Ventas'],
    locationmode='USA-states',
    colorscale='Cividis',
    colorbar=dict(y=0.4),
    colorbar_title='Ventas por Estado',
    showscale=True
), row=1, col=2)

# Añadir burbujas para mostrar ventas por ciudad
fig.add_trace(go.Scattergeo(
    lon=ventas_por_ciudad['Longitud'],
    lat=ventas_por_ciudad['Latitud'],
    text=ventas_por_ciudad['Ciudad-Estado'] + ": " + ventas_por_ciudad['Ventas'].astype(str),
    marker=dict(
        color='#56B4E9',
        size=ventas_por_ciudad['Ventas_escaladas'],
        sizemode='diameter'
    ),
    name='Ventas por Ciudad'
), row=1, col=2)

# Configurar el layout del gráfico
fig.update_layout(
    title_text='Ventas por Estado y Ciudad en Estados Unidos',
    title_x=0.5,
    geo=dict(scope='usa', domain=dict(x=[0.3, 1], y=[0, 1])),
    width=1200,
    height=600
)

fig.show()


In [ ]:
import plotly.express as px
import calendar

# Agrupar por estado y mes
ventas_por_estado_mes = df_total.groupby(['Estado', 'Mes'])['Ventas'].sum().reset_index()

# Convertir la columna 'Mes' a nombres de meses
ventas_por_estado_mes['Mes'] = ventas_por_estado_mes['Mes'].apply(lambda x: calendar.month_name[x])

# Crear el gráfico con nombres de meses
fig = px.line(ventas_por_estado_mes, x='Mes', y='Ventas', color='Estado',
              title='Variación de Ventas por Estado a lo largo de los Meses',
              labels={'Mes': 'Mes del Año', 'Ventas': 'Ventas Totales'},
              template='plotly_dark')

fig.show()


In [ ]:
# Extraer la información de ventas por ciudad
ventas_ciudad_data = ventas_por_ciudad[['Ciudad-Estado', 'Ventas']].sort_values(by='Ventas', ascending=False)

# Extraer la información de ventas por estado
ventas_estado_data = ventas_por_estado.sort_values(by='Ventas', ascending=False)

# Convertir los DataFrames a formato CSV para análisis
ventas_ciudad_csv = ventas_ciudad_data.to_csv(index=False)
ventas_estado_csv = ventas_estado_data.to_csv(index=False)

print("Ventas por Ciudad:\n", ventas_ciudad_csv)
print("\nVentas por Estado:\n", ventas_estado_csv)

Ventas por Ciudad:
 Ciudad-Estado,Ventas
"San Francisco, CA",8262203.91
"Los Angeles, CA",5452570.8
"New York City, NY",4664317.43
"Boston, MA",3661642.0100000002
"Atlanta, GA",2795498.58
"Dallas, TX",2767975.4
"Seattle, WA",2747755.48
"Portland, OR",1870732.34
"Austin, TX",1819581.75
"Portland, ME",449758.27


Ventas por Estado:
 Estado,Ventas
CA,13714774.71
NY,4664317.43
TX,4587557.15
MA,3661642.0100000002
GA,2795498.58
WA,2747755.48
OR,1870732.34
ME,449758.27



In [ ]:
import calendar
import pandas as pd

# Suponiendo que ya tienes el DataFrame df_total cargado

# Agrupar por estado y mes
ventas_por_estado_mes = df_total.groupby(['Estado', 'Mes'])['Ventas'].sum().reset_index()

# Convertir la columna 'Mes' a nombres de meses
ventas_por_estado_mes['Mes'] = ventas_por_estado_mes['Mes'].apply(lambda x: calendar.month_name[x])

# Pivotear el DataFrame para tener meses como columnas y estados como filas
ventas_pivot = ventas_por_estado_mes.pivot(index='Estado', columns='Mes', values='Ventas')

# Mostrar el DataFrame pivotado
print(ventas_pivot)


Mes          April     August    December   February    January        July  \
Estado                                                                        
CA      1362690.21  884671.60  1790646.11  889134.00  724190.23  1037216.40   
GA       284448.91  169267.66   360899.26  176470.30  149159.54   211766.47   
MA       353807.11  239275.26   509599.16  214808.36  201088.49   291497.14   
ME        42536.49   35996.60    51966.07   29845.49   22708.80    32421.14   
NY       449447.75  303183.46   646770.83  305372.26  260591.29   355716.10   
OR       197441.63  116881.14   251748.04  119606.37   92276.76   143994.49   
TX       424044.07  305477.07   614495.94  295455.39  231549.57   362650.10   
WA       276254.07  189715.09   387317.93  171330.25  140692.05   212513.92   

Mes           June       March         May    November     October  September  
Estado                                                                         
CA      1064705.41  1123656.33  1276368.70  12646

<h1>Análisis:</h1>
El análisis se centra en las ventas registradas en diferentes ciudades y estados de Estados Unidos. A partir de los datos proporcionados, se pueden identificar las ciudades y estados con las mayores ventas, así como las tendencias generales en la distribución de las ventas.

- **Ventas por Ciudad:** San Francisco, CA lidera la lista con ventas de aproximadamente 8,262,203.91, seguida por Los Angeles, CA con ventas de 5,452,570.80 y New York City, NY con 4,664,317.43.
Las ciudades de Boston, MA y Atlanta, GA también muestran ventas significativas, superando los 2.5 millones.
Portland, ME tiene las ventas más bajas entre las ciudades listadas, con un total de 449,758.27.

- **Ventas por Estado:**
California (CA) encabeza la lista con un total de ventas de 13,714,774.71, lo que es coherente con el hecho de que dos de sus ciudades, San Francisco y Los Ángeles, están entre las principales ciudades en ventas.
Nueva York (NY) y Texas (TX) siguen en la lista con ventas de 4,664,317.43 y 4,587,557.15 respectivamente.
Maine (ME) registra las ventas más bajas entre los estados listados con 449,758.27.

<h1>Respuestas a las preguntas:</h1>

- **¿En qué ciudades se han registrado las mayores ventas?**

 - Las tres ciudades con las mayores ventas son San Francisco, CA, Los Angeles, CA y New York City, NY.

- **¿Cómo se comparan las ventas en diferentes estados o regiones?**

 - California (CA) domina claramente en ventas, seguido por Nueva York (NY) y Texas (TX). Los demás estados tienen ventas significativamente más bajas en comparación.

- **¿Existe variación en las ventas por estado a lo largo de los meses?**

 - Sí, hay variaciones notables en las ventas por estado a lo largo de los meses. Por ejemplo, California (CA) muestra un aumento en abril, una ligera disminución y luego un aumento significativo hacia fin de año, especialmente en diciembre. Esta tendencia es un patrón que se refleja en todos los estados, con un pico en diciembre.

<h1>Conclusión:</h1>

> Las ventas varían considerablemente entre diferentes ciudades y estados. Es vital para las empresas comprender estas diferencias geográficas para adaptar sus estrategias de marketing y distribución. Las tendencias sugieren una mayor actividad comercial hacia el final del año, lo que las empresas deben tener en cuenta al planificar sus campañas.

### **4. Análisis del producto más vendido:**
- ¿Cuál es el producto más vendido en general y en cada mes?
- ¿Qué factores crees que han contribuido al éxito de ese producto en particular?

In [ ]:
import plotly.graph_objects as go
import plotly.express as px

# Agrupar por producto y sumar las ventas
ventas_por_producto = df_total.groupby('Producto')['Ventas'].sum().reset_index()

# Productos más vendidos en general (top 3)
productos_top = ventas_por_producto.sort_values(by='Ventas', ascending=False).iloc[:3]

# Lista de los tres productos más vendidos
top_productos_lista = productos_top['Producto'].tolist()

# Filtrar el dataframe para obtener solo los datos de los tres productos más vendidos
ventas_top_productos_mes = df_total[df_total['Producto'].isin(top_productos_lista)]

# Agrupar por producto y mes
ventas_por_producto_mes = ventas_top_productos_mes.groupby(['Producto', 'Mes'])['Ventas'].sum().reset_index()

# Gráfico de barras para el producto más vendido en general
fig1 = px.bar(ventas_por_producto, x='Producto', y='Ventas', title=f"Productos más vendidos en general", color='Ventas')

# Gráfico de líneas para los productos más vendidos por mes
fig2 = px.line(ventas_por_producto_mes, x='Mes', y='Ventas', color='Producto', title="Productos más vendidos por mes")

# Mostrar gráficos
fig1.show()
fig2.show()


In [ ]:
import plotly.graph_objects as go
import plotly.express as px

# Agrupar por producto y sumar las unidades vendidas
unidades_por_producto = df_total.groupby('Producto')['Cantidad Pedida'].sum().reset_index()

# Productos más vendidos en general en términos de unidades (top 3)
productos_top_unidades = unidades_por_producto.sort_values(by='Cantidad Pedida', ascending=False).iloc[:3]

# Lista de los tres productos más vendidos en términos de unidades
top_productos_unidades_lista = productos_top_unidades['Producto'].tolist()

# Filtrar el dataframe para obtener solo los datos de los tres productos más vendidos en términos de unidades
unidades_top_productos_mes = df_total[df_total['Producto'].isin(top_productos_unidades_lista)]

# Agrupar por producto y mes
unidades_por_producto_mes = unidades_top_productos_mes.groupby(['Producto', 'Mes'])['Cantidad Pedida'].sum().reset_index()

# Gráfico de barras para los productos más vendidos en general en términos de unidades
fig1 = px.bar(unidades_por_producto, x='Producto', y='Cantidad Pedida', title=f"Productos más vendidos en general (unidades)", color='Cantidad Pedida')

# Gráfico de líneas para los productos más vendidos por mes en términos de unidades
fig2 = px.line(unidades_por_producto_mes, x='Mes', y='Cantidad Pedida', color='Producto', title="Productos más vendidos por mes (unidades)")

# Mostrar gráficos
fig1.show()
fig2.show()


In [ ]:
# Agrupar por producto y sumar las ventas
ventas_por_producto = df_total.groupby('Producto')['Ventas'].sum().reset_index()

# Producto más vendido en general en términos de ventas
producto_top_ventas = ventas_por_producto.sort_values(by='Ventas', ascending=False).iloc[0]

print("Producto más vendido en general (ventas):")
print(producto_top_ventas)

# Agrupar por producto y sumar las unidades vendidas
unidades_por_producto = df_total.groupby('Producto')['Cantidad Pedida'].sum().reset_index()

# Producto más vendido en general en términos de unidades
producto_top_unidades = unidades_por_producto.sort_values(by='Cantidad Pedida', ascending=False).iloc[0]

print("\nProducto más vendido en general (unidades):")
print(producto_top_unidades)

# Agrupar por producto y mes y sumar las ventas
ventas_por_producto_mes = df_total.groupby(['Producto', 'Mes'])['Ventas'].sum().reset_index()

# Producto más vendido por mes en términos de ventas
productos_top_mes_ventas = ventas_por_producto_mes.loc[ventas_por_producto_mes.groupby('Mes')['Ventas'].idxmax()]

print("\nProducto más vendido por mes (ventas):")
print(productos_top_mes_ventas)

# Agrupar por producto y mes y sumar las unidades vendidas
unidades_por_producto_mes = df_total.groupby(['Producto', 'Mes'])['Cantidad Pedida'].sum().reset_index()

# Producto más vendido por mes en términos de unidades
productos_top_mes_unidades = unidades_por_producto_mes.loc[unidades_por_producto_mes.groupby('Mes')['Cantidad Pedida'].idxmax()]

print("\nProducto más vendido por mes (unidades):")
print(productos_top_mes_unidades)



Producto más vendido en general (ventas):
Producto    Macbook Pro Laptop
Ventas               8037600.0
Name: 13, dtype: object

Producto más vendido en general (unidades):
Producto           AAA Batteries (4-pack)
Cantidad Pedida                     31017
Name: 5, dtype: object

Producto más vendido por mes (ventas):
               Producto  Mes     Ventas
156  Macbook Pro Laptop    1   399500.0
157  Macbook Pro Laptop    2   469200.0
158  Macbook Pro Laptop    3   644300.0
159  Macbook Pro Laptop    4   771800.0
160  Macbook Pro Laptop    5   790500.0
161  Macbook Pro Laptop    6   605200.0
162  Macbook Pro Laptop    7   625600.0
163  Macbook Pro Laptop    8   508300.0
164  Macbook Pro Laptop    9   489600.0
165  Macbook Pro Laptop   10   892500.0
166  Macbook Pro Laptop   11   748000.0
167  Macbook Pro Laptop   12  1093100.0

Producto más vendido por mes (unidades):
                  Producto  Mes  Cantidad Pedida
60  AAA Batteries (4-pack)    1             1600
61  AAA Batteries (4

<h1>Análisis: </h1>

- **¿Cuál es el producto más vendido en general y en cada mes?**

 - **En términos de ventas totales:** El producto más vendido en general es el "Macbook Pro Laptop" con ventas totales de 8,037,600.0. Además, el "Macbook Pro Laptop" también lidera las ventas en cada mes del año, lo que indica una demanda constante y fuerte a lo largo del año.

 - **En términos de unidades vendidas:** El producto más vendido en general es "AAA Batteries (4-pack)" con un total de 31,017 unidades vendidas. Al observar las ventas mensuales, las "AAA Batteries (4-pack)" también son el producto más vendido en términos de unidades en cada mes del año.

- **¿Qué factores crees que han contribuido al éxito de ese producto en particular?**

 - Macbook Pro Laptop:

   - Calidad y Marca: Apple, la empresa detrás del Macbook Pro, es conocida por producir productos de alta calidad y tiene una base de clientes leales. La reputación de la marca y la calidad del producto pueden haber contribuido significativamente a las altas ventas.
   - Demanda de Tecnología: En la era digital actual, las laptops son esenciales para muchas actividades, desde el trabajo hasta el entretenimiento. La demanda constante de tecnología de vanguardia puede haber impulsado las ventas del Macbook Pro.
   - Marketing y Publicidad: Apple invierte significativamente en marketing y publicidad, lo que podría haber aumentado la visibilidad y la demanda del producto.

  - AAA Batteries (4-pack):

   - Uso Generalizado: Las baterías AAA son comunes y se utilizan en una amplia variedad de dispositivos, desde controles remotos hasta juguetes. La necesidad constante de baterías en los hogares puede haber impulsado las ventas.
   - Precio: A menudo, las baterías AAA son productos de bajo costo en comparación con otros productos electrónicos. La relación calidad-precio puede haber atraído a más consumidores.
   - Necesidad de Reposición: A diferencia de muchos otros productos, las baterías se agotan y requieren reemplazo regularmente, lo que puede haber llevado a compras repetidas y, por lo tanto, a un mayor volumen de ventas.

<h1>Conclusión:</h1>

> El análisis de los productos más vendidos revela la importancia de comprender las necesidades y preferencias del cliente. Mientras que productos de alta gama como el Macbook Pro Laptop dominan en términos de ingresos, productos esenciales y de uso diario como las baterías AAA lideran en términos de unidades vendidas. Las empresas deben considerar tanto el valor como el volumen al planificar estrategias de producto y marketing.

### **5. Tendencia ventas:**
- ¿Existe alguna tendencia o patrón en las ventas que se repita a lo largo de los días en los diferentes meses?
- ¿Cómo varían las ventas a lo largo de los diferentes días de la semana?
- ¿Se observa alguna diferencia significativa en las ventas entre los días laborables y los fines de semana?

In [ ]:
import calendar
import plotly.graph_objects as go

# Convertir números de mes a nombres de mes
df_total['Nombre del Mes'] = df_total['Mes'].apply(lambda x: calendar.month_name[x])

# Convertir números de día a nombres de día de la semana
df_total['Nombre del Día'] = df_total['Fecha de Pedido'].dt.day_name()

ventas_por_mes_dia = df_total.groupby(['Nombre del Mes', 'Nombre del Día']).sum()['Ventas'].reset_index()

# Crear una figura vacía
fig = go.Figure()

# Orden de los días de la semana
dias_ordenados = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

# Para cada mes, agregar una línea al gráfico que represente las ventas por día de la semana
for mes in ventas_por_mes_dia['Nombre del Mes'].unique():
    df_mes = ventas_por_mes_dia[ventas_por_mes_dia['Nombre del Mes'] == mes]
    fig.add_trace(go.Scatter(x=dias_ordenados,
                             y=[df_mes[df_mes['Nombre del Día'] == dia]['Ventas'].values[0] if dia in df_mes['Nombre del Día'].values else 0 for dia in dias_ordenados],
                             mode='lines+markers',
                             name=mes))

# Configurar el layout del gráfico con el tema oscuro
fig.update_layout(template="plotly_dark",
                  title='Ventas por Día de la Semana en Diferentes Meses',
                  xaxis_title='Día de la Semana',
                  yaxis_title='Ventas ($)',
                  xaxis={'categoryorder':'array', 'categoryarray':dias_ordenados})

# Mostrar el gráfico
fig.show()


<ipython-input-61-3e61aa073e77>:10: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



In [ ]:
import calendar
import plotly.graph_objects as go

# Convertir números de mes a nombres de mes
df_total['Nombre del Mes'] = df_total['Mes'].apply(lambda x: calendar.month_name[x])

# Convertir números de día a nombres de día de la semana
df_total['Nombre del Día'] = df_total['Fecha de Pedido'].dt.day_name()

ventas_por_mes_dia = df_total.groupby(['Nombre del Mes', 'Nombre del Día']).sum()['Ventas'].reset_index()

# Orden de los días de la semana
dias_ordenados = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

# Crear una figura vacía
fig = go.Figure()

# Para cada mes, agregar una traza de barras al gráfico que represente las ventas por día de la semana
for mes in ventas_por_mes_dia['Nombre del Mes'].unique():
    df_mes = ventas_por_mes_dia[ventas_por_mes_dia['Nombre del Mes'] == mes]
    fig.add_trace(go.Bar(x=dias_ordenados,
                         y=[df_mes[df_mes['Nombre del Día'] == dia]['Ventas'].values[0] if dia in df_mes['Nombre del Día'].values else 0 for dia in dias_ordenados],
                         name=mes))

# Configurar el layout del gráfico con el tema oscuro y barras apiladas
fig.update_layout(template="plotly_dark",
                  title='Ventas por Día de la Semana en Diferentes Meses',
                  xaxis_title='Día de la Semana',
                  yaxis_title='Ventas ($)',
                  xaxis={'categoryorder':'array', 'categoryarray':dias_ordenados},
                  barmode='stack')

# Mostrar el gráfico
fig.show()


<ipython-input-62-293ef89cc34f>:10: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



In [ ]:
# Gráfico de líneas: Ventas por Día de la Semana en Diferentes Meses
# Crear un diccionario para almacenar la descripción de las ventas por día de la semana
descripcion_ventas_dia = {}

# Para cada día de la semana
for dia in dias_ordenados:
    df_dia = ventas_por_mes_dia[ventas_por_mes_dia['Nombre del Día'] == dia]
    mes_max_ventas = df_dia[df_dia['Ventas'] == df_dia['Ventas'].max()]['Nombre del Mes'].values[0]
    mes_min_ventas = df_dia[df_dia['Ventas'] == df_dia['Ventas'].min()]['Nombre del Mes'].values[0]
    max_ventas = df_dia['Ventas'].max()
    min_ventas = df_dia['Ventas'].min()
    promedio_ventas = df_dia['Ventas'].mean()
    mediana_ventas = df_dia['Ventas'].median()

    descripcion_ventas_dia[dia] = f"El {dia}, el mes con las ventas más altas fue {mes_max_ventas} con ${max_ventas:.2f}, el mes con las ventas más bajas fue {mes_min_ventas} con ${min_ventas:.2f}. El promedio de ventas fue de ${promedio_ventas:.2f} y la mediana fue de ${mediana_ventas:.2f}."

# Imprimir la descripción
for dia, desc in descripcion_ventas_dia.items():
    print(desc)
    print("-----")

# Gráfico de barras apiladas: Ventas por Día de la Semana en Diferentes Meses
# Crear un diccionario para almacenar la descripción de la contribución de ventas por día de la semana
descripcion_contribucion_dia = {}

# Para cada día de la semana
for dia in dias_ordenados:
    df_dia = ventas_por_mes_dia[ventas_por_mes_dia['Nombre del Día'] == dia]
    mes_max_contribucion = df_dia[df_dia['Ventas'] == df_dia['Ventas'].max()]['Nombre del Mes'].values[0]
    mes_min_contribucion = df_dia[df_dia['Ventas'] == df_dia['Ventas'].min()]['Nombre del Mes'].values[0]
    max_contribucion = df_dia['Ventas'].max()
    min_contribucion = df_dia['Ventas'].min()

    descripcion_contribucion_dia[dia] = f"El {dia}, el mes que contribuyó con la mayor parte de las ventas fue {mes_max_contribucion} con ${max_contribucion:.2f} y el mes que contribuyó con la menor parte fue {mes_min_contribucion} con ${min_contribucion:.2f}."

# Imprimir la descripción
for dia, desc in descripcion_contribucion_dia.items():
    print(desc)
    print("-----")


El Monday, el mes con las ventas más altas fue December con $763665.71, el mes con las ventas más bajas fue January con $222134.67. El promedio de ventas fue de $406943.89 y la mediana fue de $381478.32.
-----
El Tuesday, el mes con las ventas más altas fue December con $747111.03, el mes con las ventas más bajas fue September con $265523.63. El promedio de ventas fue de $423996.40 y la mediana fue de $388973.54.
-----
El Wednesday, el mes con las ventas más altas fue December con $619961.91, el mes con las ventas más bajas fue August con $286802.72. El promedio de ventas fue de $415735.19 y la mediana fue de $394298.09.
-----
El Thursday, el mes con las ventas más altas fue October con $596239.28, el mes con las ventas más bajas fue January con $277091.86. El promedio de ventas fue de $403288.76 y la mediana fue de $360516.64.
-----
El Friday, el mes con las ventas más altas fue December con $592051.96, el mes con las ventas más bajas fue January con $246617.98. El promedio de ventas 

In [ ]:
# Gráfico de líneas: Ventas por Día de la Semana en Diferentes Meses
# Crear un diccionario para almacenar la descripción de las ventas por día de la semana
descripcion_ventas_dia = {}

# Orden de los meses
meses_ordenados = list(calendar.month_name[1:])

# Para cada día de la semana
for dia in dias_ordenados:
    df_dia = ventas_por_mes_dia[ventas_por_mes_dia['Nombre del Día'] == dia].sort_values(by='Nombre del Mes')
    df_dia['Diferencia'] = df_dia['Ventas'].diff()

    mes_max_ventas = df_dia[df_dia['Ventas'] == df_dia['Ventas'].max()]['Nombre del Mes'].values[0]
    mes_min_ventas = df_dia[df_dia['Ventas'] == df_dia['Ventas'].min()]['Nombre del Mes'].values[0]
    mes_subiendo = df_dia[df_dia['Diferencia'] > 0]['Nombre del Mes'].tolist()
    mes_bajando = df_dia[df_dia['Diferencia'] < 0]['Nombre del Mes'].tolist()
    mes_estable = df_dia[abs(df_dia['Diferencia']) < df_dia['Diferencia'].std()]['Nombre del Mes'].tolist()

    descripcion_ventas_dia[dia] = f"El {dia}, el mes con las ventas más altas fue {mes_max_ventas} y el mes con las ventas más bajas fue {mes_min_ventas}. Las ventas comenzaron a subir en los meses: {', '.join(mes_subiendo)}. Las ventas comenzaron a bajar en los meses: {', '.join(mes_bajando)}. Las ventas fueron más estables en los meses: {', '.join(mes_estable)}."

# Imprimir la descripción
for dia, desc in descripcion_ventas_dia.items():
    print(desc)
    print("-----")


El Monday, el mes con las ventas más altas fue December y el mes con las ventas más bajas fue January. Las ventas comenzaron a subir en los meses: December, July, March, May, November, October. Las ventas comenzaron a bajar en los meses: August, February, January, June, September. Las ventas fueron más estables en los meses: August, January, July, June, March, May, November, October, September.
-----
El Tuesday, el mes con las ventas más altas fue December y el mes con las ventas más bajas fue September. Las ventas comenzaron a subir en los meses: December, July, March, May, November, October. Las ventas comenzaron a bajar en los meses: August, February, January, June, September. Las ventas fueron más estables en los meses: January, July, June, March, May, November, October.
-----
El Wednesday, el mes con las ventas más altas fue December y el mes con las ventas más bajas fue August. Las ventas comenzaron a subir en los meses: December, July, March, May, October. Las ventas comenzaron 

<h1>Análisis: </h1>

- **¿Existe alguna tendencia o patrón en las ventas que se repita a lo largo de los días en los diferentes meses?**

 - **Tendencia General:** Diciembre es consistentemente el mes con las ventas más altas en todos los días de la semana, lo que sugiere un fuerte impulso de compra durante la temporada navideña.

 - **Inicio de Semana (Lunes a Miércoles):** Las ventas tienden a ser más estables con ligeras fluctuaciones. Sin embargo, se observa un aumento en las ventas en meses como julio y octubre, además de diciembre.

 - **Fin de Semana (Viernes a Domingo):** Las ventas suelen ser más altas, especialmente los domingos. Aunque diciembre lidera, otros meses como julio y octubre también muestran un buen rendimiento en ventas.

 - **Meses con Bajas Ventas:** Enero y septiembre suelen tener las ventas más bajas en la mayoría de los días de la semana, lo que podría estar relacionado con el fin de la temporada navideña y el inicio del año escolar, respectivamente.

- **¿Cómo varían las ventas a lo largo de los diferentes días de la semana?**

 - Las ventas tienden a ser más altas durante el fin de semana, especialmente los domingos, lo que podría estar relacionado con el tiempo libre que tienen las personas para comprar. Sin embargo, es importante destacar que diciembre es consistentemente el mes con las ventas más altas, independientemente del día de la semana, lo que indica una fuerte demanda durante la temporada navideña.

- **¿Se observa alguna diferencia significativa en las ventas entre los días laborables y los fines de semana?**

 - Sí, se observa que las ventas tienden a ser más altas durante los fines de semana en comparación con los días laborables. Esto podría deberse a que las personas tienen más tiempo libre para comprar durante el fin de semana. Sin embargo, es importante tener en cuenta que las festividades y las promociones pueden influir en estas tendencias.

<h1>Conclusión:</h1>

> El análisis de las ventas por día de la semana en diferentes meses revela patrones interesantes. En resumen, hay una clara tendencia de aumento de ventas en diciembre en todos los días de la semana. Además, se observan patrones de compra específicos en ciertos meses dependiendo del día de la semana. Las empresas deben tener en cuenta estas tendencias al planificar sus estrategias.


### **6. Impacto de eventos especiales en las ventas:**

- ¿Se ha observado algún aumento o disminución significativa en las ventas en días cercanos a eventos especiales, como días festivos?

- ¿Qué eventos específicos han tenido un impacto notable en el comportamiento de
las ventas y cómo se manifestó ese impacto?

In [ ]:
import pandas as pd

# Datos de eventos externos
eventos_externos = [
    ["2019-09-10", "Evento de Apple - Lanzamiento de productos", "10:00:00", 9, 10, 2019],
    ["2019-07-09", "Lanzamiento de MacBook Pro", "XX:XX:XX", 7, 9, 2019],
]

# Crear un DataFrame a partir de los datos de eventos externos
df_eventos_externos = pd.DataFrame(eventos_externos, columns=["Fecha", "Evento", "Hora", "Mes", "Día", "Año"])

# Guardar el DataFrame en un archivo CSV
df_eventos_externos.to_csv("eventos_externos.csv", index=False)

print("Archivo CSV de eventos externos modificado y creado exitosamente.")


Archivo CSV de eventos externos modificado y creado exitosamente.


In [ ]:
# Convertir la columna 'Fecha de Pedido' en objetos DateTime
df_total['Fecha de Pedido'] = pd.to_datetime(df_total['Fecha de Pedido'])
df_eventos_externos['Fecha'] = pd.to_datetime(df_eventos_externos['Fecha'])

# Realizar un merge en función de la columna de fecha
df_combinado = df_total.merge(df_eventos_externos, left_on='Fecha de Pedido', right_on='Fecha', how='left')


In [ ]:
import pandas as pd
import plotly.graph_objects as go

# Datos proporcionados
# ... (aquí irían tus datos de df_total y df_eventos_externos) ...

# Crear una columna de mes-año para agrupar las ventas
df_total['Mes-Año'] = df_total['Fecha de Pedido'].dt.strftime('%Y-%m')

# Identificar los productos más vendidos
top_productos = df_total.groupby('Producto').sum()['Ventas'].nlargest(5).index.tolist()

# Filtrar las ventas mensuales para esos productos
ventas_mensuales = df_total[df_total['Producto'].isin(top_productos)].groupby(['Mes-Año', 'Producto']).sum()['Ventas'].reset_index()

# Crear el gráfico con Plotly
fig = go.Figure()

# Añadir la línea de ventas mensuales para cada producto
for producto in top_productos:
    df_producto = ventas_mensuales[ventas_mensuales['Producto'] == producto]
    fig.add_trace(go.Scatter(x=df_producto['Mes-Año'], y=df_producto['Ventas'], mode='lines+markers', name=producto))

# Añadir líneas verticales para los eventos externos
for _, row in df_eventos_externos.iterrows():
    mes_año_evento = row['Fecha'].strftime('%Y-%m')
    fig.add_shape(
        type="line",
        x0=mes_año_evento,
        x1=mes_año_evento,
        y0=0,
        y1=ventas_mensuales['Ventas'].max(),
        line=dict(color="Red", width=2, dash="dot")
    )
    fig.add_annotation(
        x=mes_año_evento,
        y=ventas_mensuales['Ventas'].max(),
        text=row['Evento'],
        showarrow=True,
        arrowhead=4,
        ax=0,
        ay=-40
    )

fig.update_layout(title='Ventas mensuales de los productos más vendidos con eventos externos', xaxis_title='Mes', yaxis_title='Ventas')
fig.show()


<ipython-input-67-d81a9cf0b764>:11: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.

<ipython-input-67-d81a9cf0b764>:14: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



In [ ]:
import pandas as pd

# Datos de eventos externos
eventos_externos = [
    ["2019-01-19", "Tormenta de Nieve en el Noreste de EE. UU.", "XX:XX:XX", 1, 19, 2019],
    ["2019-09-14", "Huracán Dorian - Acercamiento a EE. UU.", "20:00:00", 9, 14, 2019],
]

df_total['Fecha'] = df_total['Fecha de Pedido']

# Crear un DataFrame a partir de los datos de eventos externos
df_eventos_externos = pd.DataFrame(eventos_externos, columns=["Fecha", "Evento", "Hora", "Mes", "Día", "Año"])

# Convertir la columna 'Fecha' en objetos DateTime
df_total['Fecha'] = pd.to_datetime(df_total['Fecha'])
df_eventos_externos['Fecha'] = pd.to_datetime(df_eventos_externos['Fecha'])

# Realizar un merge en función de la columna de fecha
df_combinado = df_total.merge(df_eventos_externos, on='Fecha', how='left')

In [ ]:
import pandas as pd
import plotly.graph_objects as go

# Crear una columna de mes-año para agrupar las ventas
df_total['Mes-Año'] = df_total['Fecha'].dt.strftime('%Y-%m')

# Identificar los productos más pedidos
top_productos_pedidos = df_total.groupby('Producto').sum()['Cantidad Pedida'].nlargest(5).index.tolist()

# Filtrar las cantidades pedidas mensuales para esos productos
pedidos_mensuales = df_total[df_total['Producto'].isin(top_productos_pedidos)].groupby(['Mes-Año', 'Producto']).sum()['Cantidad Pedida'].reset_index()

# Crear el gráfico con Plotly
fig = go.Figure()

# Añadir la línea de cantidades pedidas mensuales para cada producto
for producto in top_productos_pedidos:
    df_producto = pedidos_mensuales[pedidos_mensuales['Producto'] == producto]
    fig.add_trace(go.Scatter(x=df_producto['Mes-Año'], y=df_producto['Cantidad Pedida'], mode='lines+markers', name=producto))

# Añadir líneas verticales para los eventos externos
for _, row in df_eventos_externos.iterrows():
    mes_año_evento = row['Fecha'].strftime('%Y-%m')
    fig.add_shape(
        type="line",
        x0=mes_año_evento,
        x1=mes_año_evento,
        y0=0,
        y1=pedidos_mensuales['Cantidad Pedida'].max(),
        line=dict(color="Red", width=2, dash="dot")
    )
    fig.add_annotation(
        x=mes_año_evento,
        y=pedidos_mensuales['Cantidad Pedida'].max(),
        text=row['Evento'],
        showarrow=True,
        arrowhead=4,
        ax=0,
        ay=-40
    )

fig.update_layout(title='Cantidades pedidas mensuales de los productos más solicitados con eventos externos', xaxis_title='Mes', yaxis_title='Cantidad Pedida')
fig.show()


<ipython-input-69-4333636211ae>:8: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.

<ipython-input-69-4333636211ae>:11: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



<h1>Impacto de eventos especiales en las ventas:</h1>

- **¿Se ha observado algún aumento o disminución significativa en las ventas en días cercanos a eventos especiales, como días festivos?**

 - Sí, se ha observado un aumento significativo en las ventas en días cercanos a eventos especiales. En particular, después del evento de Apple en septiembre, las ventas de sus productos se dispararon en los meses siguientes. De manera similar, después del acercamiento del huracán Dorian a EE. UU. en septiembre, las ventas de baterías AAA también aumentaron notablemente en los meses subsiguientes. Si bien es posible que las ventas también estén influenciadas por las festividades de fin de año, la coincidencia temporal con estos eventos sugiere una relación directa.

- **¿Qué eventos específicos han tenido un impacto notable en el comportamiento de las ventas y cómo se manifestó ese impacto?**

 - **Evento de Apple:** Tras el lanzamiento de nuevos productos en el evento de Apple en septiembre, se observó un aumento significativo en las ventas de estos productos en los meses siguientes. Esto puede atribuirse a la anticipación y el interés generado por el evento, lo que llevó a un mayor número de consumidores a adquirir los nuevos productos lanzados.

 - **Acercamiento del huracán Dorian a EE. UU.:** Después de la noticia del acercamiento del huracán Dorian a EE. UU., hubo un aumento notable en las ventas de baterías AAA. Esto puede estar relacionado con la preparación y precaución de las personas ante un evento natural de esta magnitud, donde las baterías son esenciales para linternas y otros dispositivos en caso de cortes de energía.

<h1>Conclusión:</h1>

> En ambos casos, los eventos externos tuvieron un impacto directo en el comportamiento de compra de los consumidores, lo que llevó a un aumento en las ventas de productos específicos relacionados con esos eventos. Es esencial considerar estos factores al analizar las tendencias de ventas y al planificar estrategias de marketing y suministro en el futuro.


In [ ]:
df_eventos_externos
df_total
df_combinado

ID de Pedido                    Producto  Cantidad Pedida  \
0            176558        USB-C Charging Cable                2   
1            176559  Bose SoundSport Headphones                1   
2            176560                Google Phone                1   
3            176560            Wired Headphones                1   
4            176561            Wired Headphones                1   
...             ...                         ...              ...   
185945       259353      AAA Batteries (4-pack)                3   
185946       259354                      iPhone                1   
185947       259355                      iPhone                1   
185948       259356      34in Ultrawide Monitor                1   
185949       259357        USB-C Charging Cable                1   

        Precio Unitario     Fecha de Pedido  \
0                 11.95 2019-04-19 08:46:00   
1                 99.99 2019-04-07 22:30:00   
2                600.00 2019-04-12 14:38:00   
3                 11.99 2019-04-12 14:38:00   
4                 11.99 2019-04-30 09:27:00   
...                 ...                 ...   
185945             2.99 2019-09-17 20:56:00   
185946           700.00 2019-09-01 16:00:00   
185947           700.00 2019-09-23 07:39:00   
185948           379.99 2019-09-19 17:30:00   
185949            11.95 2019-09-30 00:18:00   

                             Dirección de Envio  Año_x  Mes_x  Dia  Hora_x  \
0                  917 1st St, Dallas, TX 75001   2019      4   19       8   
1             682 Chestnut St, Boston, MA 02215   2019      4    7      22   
2          669 Spruce St, Los Angeles, CA 90001   2019      4   12      14   
3          669 Spruce St, Los Angeles, CA 90001   2019      4   12      14   
4             333 8th St, Los Angeles, CA 90001   2019      4   30       9   
...                                         ...    ...    ...  ...     ...   
185945   840 Highland St, Los Angeles, CA 90001   2019      9   17      20   
185946  216 Dogwood St, San Francisco, CA 94016   2019      9    1      16   
185947     220 12th St, San Francisco, CA 94016   2019      9   23       7   
185948   511 Forest St, San Francisco, CA 94016   2019      9   19      17   
185949   250 Meadow St, San Francisco, CA 94016   2019      9   30       0   

        ...      Ciudad-Estado Nombre del Mes  Nombre del Día  Mes-Año  \
0       ...         Dallas, TX          April          Friday  2019-04   
1       ...         Boston, MA          April          Sunday  2019-04   
2       ...    Los Angeles, CA          April          Friday  2019-04   
3       ...    Los Angeles, CA          April          Friday  2019-04   
4       ...    Los Angeles, CA          April         Tuesday  2019-04   
...     ...                ...            ...             ...      ...   
185945  ...    Los Angeles, CA      September         Tuesday  2019-09   
185946  ...  San Francisco, CA      September          Sunday  2019-09   
185947  ...  San Francisco, CA      September          Monday  2019-09   
185948  ...  San Francisco, CA      September        Thursday  2019-09   
185949  ...  San Francisco, CA      September          Monday  2019-09   

                     Fecha Evento Hora_y Mes_y Día Año_y  
0      2019-04-19 08:46:00    NaN    NaN   NaN NaN   NaN  
1      2019-04-07 22:30:00    NaN    NaN   NaN NaN   NaN  
2      2019-04-12 14:38:00    NaN    NaN   NaN NaN   NaN  
3      2019-04-12 14:38:00    NaN    NaN   NaN NaN   NaN  
4      2019-04-30 09:27:00    NaN    NaN   NaN NaN   NaN  
...                    ...    ...    ...   ...  ..   ...  
185945 2019-09-17 20:56:00    NaN    NaN   NaN NaN   NaN  
185946 2019-09-01 16:00:00    NaN    NaN   NaN NaN   NaN  
185947 2019-09-23 07:39:00    NaN    NaN   NaN NaN   NaN  
185948 2019-09-19 17:30:00    NaN    NaN   NaN NaN   NaN  
185949 2019-09-30 00:18:00    NaN    NaN   NaN NaN   NaN  

[185950 rows x 23 columns]

### ***Pregunta Adicional***

Además de las preguntas mencionadas anteriormente, te invito a pensar en una pregunta adicional que podría ser relevante para analizar los datos de ventas mensuales. Esta pregunta debe estar basada en la información que se proporciona en los conjuntos de datos. Puede estar relacionada con tendencias, comparaciones, patrones o cualquier otro aspecto que despierte tu interés. Trata de ser creativo/a y pensar en cómo podrías explorar aún más estos datos!

## **¿Cómo se distribuyen las ventas de los productos más vendidos a lo largo del año?**

In [ ]:
!pip install bar_chart_race

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.8/156.8 kB 1.9 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import bar_chart_race as bcr

# Agrupar por año, mes, día y producto y sumar las ventas
df_grouped = df_total.groupby(['Año', 'Mes', 'Dia', 'Producto']).sum()['Ventas'].reset_index()

# Renombrar las columnas para que sean compatibles con pd.to_datetime
df_grouped = df_grouped.rename(columns={'Año': 'year', 'Mes': 'month', 'Dia': 'day'})

# Crear la columna 'Fecha'
df_grouped['Fecha'] = pd.to_datetime(df_grouped[['year', 'month', 'day']])

# Pivotear el DataFrame para que los productos sean columnas
df_pivot = df_grouped.pivot(index='Fecha', columns='Producto', values='Ventas').fillna(0)

# Calcular las ventas acumuladas
df_cumsum = df_pivot.cumsum()

bcr.bar_chart_race(df_cumsum, n_bars=10, sort='desc', title='Ventas acumuladas de productos a lo largo del año')


In [ ]:
import pandas as pd
import bar_chart_race as bcr

# Agrupar por Fecha y Producto y sumar la Cantidad Pedida
df_grouped = df_total.groupby(['Año', 'Mes', 'Dia', 'Producto']).sum()['Cantidad Pedida'].reset_index()

# Renombrar las columnas para que sean compatibles con pd.to_datetime
df_grouped = df_grouped.rename(columns={'Año': 'year', 'Mes': 'month', 'Dia': 'day'})

# Crear la columna 'Fecha'
df_grouped['Fecha'] = pd.to_datetime(df_grouped[['year', 'month', 'day']])

# Agrupar nuevamente por Fecha y Producto y sumar la Cantidad Pedida para evitar duplicados
df_grouped_sum = df_grouped.groupby(['Fecha', 'Producto']).sum()['Cantidad Pedida'].reset_index()

# Pivotear el DataFrame para que los productos sean columnas
df_pivot = df_grouped_sum.pivot(index='Fecha', columns='Producto', values='Cantidad Pedida').fillna(0)

# Calcular la cantidad acumulada de productos pedidos
df_cumsum = df_pivot.cumsum()

# Crear el gráfico de bar chart race
bcr.bar_chart_race(df_cumsum, n_bars=10, sort='desc', title='Cantidad acumulada de productos pedidos a lo largo del año')
